moveit_core/trajectory_processing/src/ruckig_traj_smoothing.cpp

```diff
@@ -0,0 +1,169 @@
+ *   this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *******************************************************************************/
+
+/* Author: Jack Center, Wyatt Rees, Andy Zelenak */
+
+#include <algorithm>
+#include <cmath>
+#include <Eigen/Geometry>
+#include <limits>
+#include <moveit/trajectory_processing/ruckig_traj_smoothing.h>
+#include <rclcpp/rclcpp.hpp>
+#include <ruckig/ruckig.hpp>
+#include <vector>
+
+namespace trajectory_processing
+{
+static const rclcpp::Logger LOGGER = rclcpp::get_logger("moveit_trajectory_processing.ruckig_traj_smoothing");
+constexpr double DEFAULT_MAX_VELOCITY = 5;       // rad/s
+constexpr double DEFAULT_MAX_ACCELERATION = 10;  // rad/s^2
+constexpr double DEFAULT_MAX_JERK = 20;          // rad/s^3
+
+bool RuckigSmoothing::applySmoothing(robot_trajectory::RobotTrajectory& trajectory,
+                                     const double max_velocity_scaling_factor,
+                                     const double max_acceleration_scaling_factor) const
+{
+  const moveit::core::JointModelGroup* group = trajectory.getGroup();
+  if (!group)
+  {
+    RCLCPP_ERROR(LOGGER, "It looks like the planner did not set the group the plan was computed for");
+    return false;
+  }
+
+  size_t num_waypoints = trajectory.getWayPointCount();
+  if (num_waypoints < 2)
+  {
+    RCLCPP_WARN_STREAM(LOGGER, "Trajectory does not have enough points to smooth with Ruckig");
+    return false;
+  }
+
+  const size_t num_dof = group->getVariableCount();
+
+  // This lib does not actually work properly when angles wrap around, so we need to unwind the path first
+  trajectory.unwind();
+
+  // Instantiate the smoother
+  double timestep = trajectory.getAverageSegmentDuration();
+  ruckig::Ruckig<0> ruckig{ num_dof, timestep };
+  ruckig::InputParameter<0> ruckig_input{ num_dof };
+  ruckig::OutputParameter<0> ruckig_output{ num_dof };
+
+  // Initialize the smoother
+  //  ruckig_input.synchronization = ruckig::Synchronization::Time;
+  //  ruckig_input.interface = ruckig::Interface::Velocity;
+  const std::vector<int>& idx = group->getVariableIndexList();
```

Comment:
same here, instead of iterating all variables, better use `group->getActiveJointModelNames()` and iterate those. Otherwise this could have odd effects with robots that have mimic joints (e.g. controller rejecting trajectories)

Reply:
Well, this bit was directly copied from TOTG. Open an issue to figure it out later?

Reply:
https://github.com/ros-planning/moveit2/issues/586